core/clients/key_flow_continuous_refresh.go

```diff
@@ -0,0 +1,126 @@
+	timeStartBeforeTokenExpiration time.Duration
+	timeBetweenContextCheck        time.Duration
+	timeBetweenTries               time.Duration
+}
+
+// Continuously refreshes the token of a key flow, retrying if the token API returns 5xx errrors. Always returns with a non-nil error.
+//
+// To terminate this routine, close the context in refresher.keyFlow.config.BackgroundTokenRefreshContext.
+func (refresher *continuousTokenRefresher) continuousRefreshToken() error {
+	expirationTimestamp, err := refresher.getAccessTokenExpirationTimestamp()
+	if err != nil {
+		return fmt.Errorf("get access token expiration timestamp: %w", err)
+	}
+	startRefreshTimestamp := expirationTimestamp.Add(-refresher.timeStartBeforeTokenExpiration)
+
+	for {
+		err = refresher.waitUntilTimestamp(startRefreshTimestamp)
+		if err != nil {
+			return err
+		}
+
+		err := refresher.keyFlow.config.BackgroundTokenRefreshContext.Err()
+		if err != nil {
+			return fmt.Errorf("check context: %w", err)
+		}
+
+		ok, err := refresher.refreshToken()
+		if err != nil {
+			return fmt.Errorf("refresh tokens: %w", err)
+		}
+		if !ok {
+			startRefreshTimestamp = startRefreshTimestamp.Add(refresher.timeBetweenTries)
+			continue
+		}
+
+		expirationTimestamp, err := refresher.getAccessTokenExpirationTimestamp()
+		if err != nil {
+			return fmt.Errorf("get access token expiration timestamp: %w", err)
+		}
+		startRefreshTimestamp = expirationTimestamp.Add(-refresher.timeStartBeforeTokenExpiration)
+	}
+}
+
+func (refresher *continuousTokenRefresher) getAccessTokenExpirationTimestamp() (*time.Time, error) {
+	token := refresher.keyFlow.token.AccessToken
+
+	// We can safely use ParseUnverified because we are not doing authentication of any kind
+	// We're just checking the expiration time
+	tokenParsed, _, err := jwt.NewParser().ParseUnverified(token, &jwt.RegisteredClaims{})
+	if err != nil {
+		return nil, fmt.Errorf("parse token: %w", err)
+	}
+	expirationTimestampNumeric, err := tokenParsed.Claims.GetExpirationTime()
+	if err != nil {
+		return nil, fmt.Errorf("get expiration timestamp: %w", err)
+	}
+	return &expirationTimestampNumeric.Time, nil
+}
+
+func (refresher *continuousTokenRefresher) waitUntilTimestamp(timestamp time.Time) error {
+	for time.Now().Before(timestamp) {
+		err := refresher.keyFlow.config.BackgroundTokenRefreshContext.Err()
+		if err != nil {
+			return fmt.Errorf("check context: %w", err)
+		}
+		time.Sleep(refresher.timeBetweenContextCheck)
+	}
+	return nil
+}
+
+// Returns:
+//   - (true, nil) if successful.
+//   - (false, nil) if not successful but should be retried.
+//   - (_, err) if not successful and shouldn't be retried.
+func (refresher *continuousTokenRefresher) refreshToken() (bool, error) {
+	err := refresher.keyFlow.createAccessTokenWithRefreshToken()
+	if err == nil {
+		return true, nil
+	}
+
+	// Should be retired if this is an API error with status code non-5xx
```

Comment:
Did you mean `Shouldn't be retried`? Or did you actually want to use `retired` as in "retire" the goroutine?

Reply:
But it says `non-5xx`, the idea is to only retry if error code is `5xx`, right?

Reply:
Ah oops, double negative